### ISSUE TITLE
isRequired for each cell of a matrix dynamic row

### ISSUE THREAD

**Issue by ashygk** [surveyjs/surveyjs#1255]
### Are you requesting a feature, reporting a bug or asking a question?

question

### What is the current behavior?

isrequired property is present for question level

### What is the expected behavior?

In a matrix dynamic, how can I assign required fields for each cell in a row?

#### Provide the test code and the tested page URL (if applicable)

Tested page URL:

Test code
`{
            pages: [
                {
                    name: "page1",
                    elements: [
                        {
                            type: "matrixdynamic",
                            name: "ComplexMedicalNeeds",
                            titleLocation: "hidden",
                            columns: [
                                {
                                    name: "Date",
                                    cellType: "datepicker",
                                    dateFormat: "mm/dd/yy",
                                    config: {
                                        showOn: "button",
                                        buttonImage:
                                        "assets/icons/date.svg",
                                        buttonImageOnly: true,
                                        buttonText: "Select date",
                                        maxDate: "+0D"
                                    },
                                    inputType: "date"
                                },
                                {
                                    name: "Narrative",
                                    cellType: "comment",
                                    maxLength: 2000,
                                    validators: [
                                        {
                                            type: "text",
                                            allowDigits: true,
                                            minLength: 0,
                                            maxLength: 2000
                                        }
                                    ]
                                },
                                {
                                    name: "Actions",
                                    visibleIf: "1 = 2"
                                }
                            ],
                            cellType: "text",
                            rowCount: 0,
                            addRowLocation: "top",
                            addRowText: "ADD COMPLEX NEED  +",
                            removeRowText: " ",
                            confirmDelete: true,
                            confirmDeleteText: "Are you sure you want to delete medical/nursing needs record?"
                        }
                    ]
                }
            ],
            showTitle: false,
            showQuestionNumbers: "off",
            questionErrorLocation: "bottom"
        }`
```
your_code_here

```

### Specify your

* browser: chrome
* browser version:
* surveyjs platform (angular or react or jquery or knockout or vue): angular 5
* surveyjs version: 1.0.35

**Comment by tsv2013** [comment]
You can use custom validation or validation on event for this purposes (https://surveyjs.io/Examples/Library/?id=validators-custom&platform=jQuery&theme=default or https://surveyjs.io/Examples/Library/?id=validators-event&platform=jQuery&theme=default)

**Comment by ashygk** [comment]
the custom validations above is for individual questions. Could you show an example for a matrix dynamic with each column having required property set?

**Comment by tsv2013** [comment]
The individual question is a matrix dynamic in this case. By the way if one can create and remove rows in a matrix - how do you plan to calculate the row is required?

**Comment by ashygk** [comment]
Below is the use case of validations that need to be performed in our case for the above mentioned json.


Click on Add complex need   button.    Remove the Date in Date field.   Click on Add complex need button again. | Please enter date.
-- | --
Click on Add complex need   button.    Enter Date. (Or Keep the default populated date)   Do not enter Narrative text.    Click on Add complex need button again. | Please enter at least one field   in Complex Medical/Nursing Needs.

**Comment by tsv2013** [comment]
Ok, thank you for the clarification

**Comment by ashygk** [comment]
@tsv2013 Thanks.

**Comment by dmitrykurmanov** [comment]
We've implemented new "onMatrixBeforeRowAdded" event and the example of using:

```js
  var json = {
    pages: [
      {
        name: "page1",
        elements: [
          {
            type: "matrixdynamic",
            name: "ComplexMedicalNeeds",
            titleLocation: "hidden",
            columns: [
              {
                name: "Date",
                cellType: "text",
                isRequired: true,
                inputType: "date"
              },
              {
                name: "Narrative",
                cellType: "comment",
                isRequired: true,
                validators: [
                  {
                    type: "text",
                    minLength: 0,
                    maxLength: 2000,
                    allowDigits: true
                  }
                ],
                maxLength: 2000
              },
              {
                name: "Actions",
                visibleIf: "1 = 2"
              }
            ],
            cellType: "text",
            rowCount: 0,
            confirmDelete: true,
            confirmDeleteText:
              "Are you sure you want to delete medical/nursing needs record?",
            addRowLocation: "top",
            addRowText: "ADD COMPLEX NEED +"
          }
        ]
      }
    ],
    showTitle: false,
    showQuestionNumbers: "off",
    questionErrorLocation: "bottom"
  };
...
survey.onMatrixBeforeRowAdded.add(function(sender, options) {
    for (
      var index = 0;
      index < options.question.generatedVisibleRows.length;
      index++
    ) {
      var rowData = (options.question.value || [])[index] || {};
      var q = options.question.generatedVisibleRows[
        index
      ].getQuestionByColumnName("Date");
      options.canAddRow = !q.hasErrors();
      if (!options.canAddRow) return;
      q = options.question.generatedVisibleRows[index].getQuestionByColumnName(
        "Narrative"
      );
      options.canAddRow = !q.hasErrors();
      if (!options.canAddRow) return;
    }
  });
```

**Comment by ashygk** [comment]
Thanks for implementing this. When will this fix be available ?

**Comment by dmitrykurmanov** [comment]
We've published 1.0.36, fix is available now.

**Comment by ashygk** [comment]
@dmitrykurmanov Thanks. I have tried the above example with the onMatrixBeforeRowAdded event and don't feel its working. Can you please help me out
